### PR TITLE
Remove simple JS->Wasm entry thunks by using metadata

### PR DESCRIPTION
**Pull request by justinmichaud**
#### 4bd20953a44ff01714fff5d40d746f8358be77d5
<pre>
Remove simple JS-&gt;Wasm entry thunks by using metadata
<a href="https://bugs.webkit.org/show_bug.cgi?id=267481">https://bugs.webkit.org/show_bug.cgi?id=267481</a>
<a href="https://rdar.apple.com/120933304">rdar://120933304</a>

Reviewed by Yusuke Suzuki.

This patch lays the groundwork for a new way to call wasm functions without JIT.
We opt in only the simplest cases for now, but in those cases, we can skip the
generated JS-&gt;WASM entry thunk. Future work can expand this to more signatures.

This also changes the way tier-up works. Today, we re-write the JS-&gt;Wasm entrypoint
call pointer. This won&apos;t work anymore, so we instead check when loading the entrypoint
for interpreted-entry callees.

This shouldn&apos;t affect the jsCallEntrypoint IC, so this is a step in the direction
of completely removing these non-ic jit thunks in the future.

* JSTests/wasm/stress/cc-int-to-int.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/SamplingProfiler.cpp:
(JSC::tierName):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::initializeCallees):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::runWithDowncast):
(JSC::Wasm::JSEntrypointInterpreterCallee::entrypointImpl const):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmCompilationMode.cpp:
(JSC::Wasm::makeString):
* Source/JavaScriptCore/wasm/WasmCompilationMode.h:
(JSC::Wasm::isOSREntry):
(JSC::Wasm::isAnyBBQ):
(JSC::Wasm::isAnyOMG):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.h:

Canonical link: <a href="https://commits.webkit.org/277566@main">https://commits.webkit.org/277566@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/85b2aa072099c40ab991536a0d9523b1082057f9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27037 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50668 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43897 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32910 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38935 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48425 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24724 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20227 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22191 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42548 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5891 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41132 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44214 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42985 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52419 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47335 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22882 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19235 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46241 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24153 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45281 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10585 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24945 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54833 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23874 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11263 "Passed tests") |

### CODE REVIEW

**Comment by justinmichaud** [comment]
```
justinmichaud@Justins-16-Inch-MBP OpenSource % run-jsc-benchmarks baseline:JSC_useIPIntWrappers=0:JSC_validateOptions=1:/Volumes/WebKit/DebugVersion/OpenSource/WebKitBuild/Debug/jsc useWrappers:JSC_validateOptions=1:/Volumes/WebKit/DebugVersion/OpenSource/WebKitBuild/Debug/jsc --microbenchmarks --benchmark=".*cc-.*" --inner 20 --outer 5
Warning: could not get svn revision information for baseline: No such file or directory - svn
Warning: could not get svn revision information for useWrappers: No such file or directory - svn
62/62                                                  -e 
Generating benchmark report at /Volumes/WebKit/DebugVersion/OpenSource/baseline_useWrappers_Microbenchmarks_Justins-16-Inch-MBP_20240222_1716_report.txt
And raw data at /Volumes/WebKit/DebugVersion/OpenSource/baseline_useWrappers_Microbenchmarks_Justins-16-Inch-MBP_20240222_1716.json

Benchmark report for Microbenchmarks on Justins-16-Inch-MBP (Mac14,6).

VMs tested:
"baseline" at /Volumes/WebKit/DebugVersion/OpenSource/WebKitBuild/Debug/jsc
    export JSC_useIPIntWrappers=0
    export JSC_validateOptions=1
"useWrappers" at /Volumes/WebKit/DebugVersion/OpenSource/WebKitBuild/Debug/jsc
    export JSC_validateOptions=1

Collected 100 samples per benchmark/VM, with 5 VM invocations per benchmark. Emitted a call to gc()
between sample measurements. Used 1 benchmark iteration per VM invocation for warm-up. Used the
jsc-specific preciseTime() function to get microsecond-level timing. Reporting benchmark execution
times with 95% confidence intervals in milliseconds.

                             baseline                useWrappers                                    

wasm-cc-int-to-int      262.3517+-0.6392     !    295.8643+-0.4344        ! definitely 1.1277x slower

<geometric>             262.3517+-0.6392     !    295.8643+-0.4344        ! definitely 1.1277x slower```

**Comment by justinmichaud** [comment]
```Collected 25 samples per benchmark/VM, with 5 VM invocations per benchmark. Emitted a call to gc()
between sample measurements. Used 1 benchmark iteration per VM invocation for warm-up. Used the
jsc-specific preciseTime() function to get microsecond-level timing. Reporting benchmark execution
times with 95% confidence intervals in milliseconds.

                             baseline                useWrappers                                    

wasm-cc-int-to-int     3749.1413+-6.4129     !   3780.3564+-11.4321       ! definitely 1.0083x slower

<geometric>            3749.1413+-6.4129     !   3780.3564+-11.4321       ! definitely 1.0083x slower```

File: Source/JavaScriptCore/runtime/OptionsList.h [hunk 1]
```diff
@@ -578,7 +578,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, useWasmIPIntEpilogueOSR, true, Normal, "Allow IPInt to tier up during function epilogues") \
     v(Bool, wasmIPIntTiersUpToBBQ, true, Normal, "Allow IPInt to tier up to BBQ") \
     v(Bool, wasmIPIntTiersUpToOMG, true, Normal, "Allow IPInt to tier up to OMG") \
-    v(Bool, useIPIntWrappers, false, Normal, "Allow IPInt to replace JIT wasm wrappers") \
+    v(Bool, useIPIntWrappers, true, Normal, "Allow IPInt to replace JIT wasm wrappers") \
```
Comment (justinmichaud) [inline Source/JavaScriptCore/runtime/OptionsList.h:581]: I will land without turning this on by default

File: Source/JavaScriptCore/wasm/WasmBBQPlan.cpp [hunk 2]
```diff
@@ -214,6 +214,43 @@ void BBQPlan::work(CompilationEffort effort)
...
+                else
+                    entrypoint = m_calleeGroup->wasmEntrypointCalleeFromFunctionIndexSpace(locker, call.functionIndexSpace).entrypoint().retagged<WasmEntryPtrTag>();
+
+                MacroAssembler::repatchNearCall(call.callLocation, CodeLocationLabel<WasmEntryPtrTag>(entrypoint));
+            }
+
+            callee->setEntrypoint(WTFMove(jsToWasmInternalFunction->entrypoint));
+            static_cast<JSEntrypointInterpreterCallee*>(jsEntrypointCallee)->setReplacement(callee.ptr());
+
+            auto result = m_jsToWasmInternalFunctions.add(m_functionIndex, std::tuple { WTFMove(callee), WTFMove(linkBuffer), WTFMove(jsToWasmInternalFunction) });
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:249]: This is a bit sketchy I guess?

File: Source/JavaScriptCore/llint/WebAssembly.asm [hunk 3]
```diff
@@ -581,9 +581,357 @@ end
...
+        _js_to_wasm_wrapper_entry_interp_afterops_validate:
+        break
+    end
+
+opcode(LoadI32)
+_js_to_wasm_wrapper_entry_interp_begin:
+    advanceSigned()
+    lshiftq 3, Scratch
+    addp cfr, Scratch
+    loadi [Scratch], Accumulator
```
Comment (Constellation) [inline Source/JavaScriptCore/llint/WebAssembly.asm:800]: Can we insert newline after each opcode to make it more readable?

File: Source/JavaScriptCore/llint/WebAssembly.asm [hunk 4]
```diff
@@ -581,9 +581,357 @@ end
...
+opcode(LoadI64)
+break
```
Comment (Constellation) [inline Source/JavaScriptCore/llint/WebAssembly.asm]: Also, let's make it indented for each opcode's body.

**Review by Constellation — APPROVED** [review]
r=me with nits.

**Pull request merged**